maro/simulator/scenarios/data_center/business_engine.py

```diff
@@ -0,0 +1,320 @@
+)
+from .events import Events
+from .physical_machine import PhysicalMachine
+from .virtual_machine import VirtualMachine
+
+metrics_desc = """
+
+energy_consumption (int): Current total energy consumption.
+success_requirements (int): Accumulative successful VM requirements until now.
+failed_requirements (int): Accumulative failed VM requirements until now.
+total_latency (int): Accumulative used buffer time until now.
+"""
+
+
+class DataCenterBusinessEngine(AbsBusinessEngine):
+    def __init__(
+        self, event_buffer: EventBuffer, topology: str, start_tick: int,
+        max_tick: int, snapshot_resolution: int, max_snapshots: int, additional_options: dict = {}
+    ):
+        super().__init__(
+            scenario_name="data_center", event_buffer=event_buffer, topology=topology, start_tick=start_tick,
+            max_tick=max_tick, snapshot_resolution=snapshot_resolution, max_snapshots=max_snapshots,
+            additional_options=additional_options
+        )
+
+        # Env metrics.
+        self._energy_consumption: int = 0
+        self._successful_requirements: int = 0
+        self._failed_requirements: int = 0
+        self._total_latency: Latency = Latency()
+
+        # Load configurations.
+        self._load_configs()
+        self._register_events()
+
+        # PMs list used for quick accessing.
+        self._init_machines()
+        # All living VMs.
+        self._live_vm: Dict[int, VirtualMachine] = {}
+        # All requirement payload of the pending decision VMs.
+        # NOTE: Need naming suggestestion.
+        self._pending_vm_req_payload: Dict[int, VmRequirementPayload] = {}
+
+        self._tick: int = 0
+        self._pending_action_vm_id: int = -1
+
+    def _load_configs(self):
+        """Load configurations."""
+
+        # Update self._config_path with current file path.
+        self.update_config_root_path(__file__)
+        with open(os.path.join(self._config_path, "config.yml")) as fp:
+            self._conf = safe_load(fp)
+
+        self._delay_duration: int = self._conf["delay_duration"]
+        self._pm_amount: int = self._conf["pm_amount"]
+        self._pm_cap_cpu: int = self._conf["pm_cap_cpu"]
+        self._pm_cap_mem: int = self._conf["pm_cap_mem"]
+
+    def _init_machines(self):
+        """Initialize the physical machines based on the config setting. The PM id starts from 0."""
+
+        self._machines: List[PhysicalMachine] = [
+            PhysicalMachine(
+                id=i,
+                cap_cpu=self._pm_cap_cpu,
+                cap_mem=self._pm_cap_mem
+            ) for i in range(self._pm_amount)
+        ]
+
+    def step(self, tick: int):
+        """Push business to next step.
+
+        Args:
+            tick (int): Current tick to process.
+        """
+        self._tick = tick
+
+        # Update all live VMs CPU utilization.
+        self._update_vm_util()
+        # Update all PM CPU utilization.
+        self._update_pm_util()
+        # TODO
```

Comment:
so after the workload updating, we need access to the log data, get a requirements list, which will be used to generate the requirement events?